docs/DesignDocs/ExtendedEnclaveInitializationData_V2.md

```diff
@@ -0,0 +1,348 @@
+| FS/GS Page (oe_sgx_td_t + tsp)  |
+| ...                             |
+| Guard Page                      |
+| Stack pages                     |
+| Guard Page                      |
+| TCS Page                        |
+| SSA (State Save Area) 0         |
+| SSA (State Save Area) 1         |
+| Guard Page                      |
+| Thread local storage            |
+| FS/GS Page (oe_sgx_td_t + tsp)  |
+
+For a regular enclave, the enclave loader "adds" a Guard Page between the Code
+and Initialized Data pages and the Heap Pages (`EADD` the first heap page at the
+linear address = linear address of the last Code/Data page + 4KB).
+
+For an EEID enclave, the enclave loader records the Guard Page location, adds
+the heap and the thread sections, and then the EEID page at the Guard Page
+location, as read-only. The Guard/EEID page location is fixed, as
+(heap_base-4KB). The EEID initialization code inside the enclave always accesses
+the EEID data structure at the fixed location.
+
+For EEID enclaves with heap/stack size and #TCS specified at enclave signing
+time, the loader only adds one extra measured page after the original base
+enclave is measured by the CPU. The `hash_state` in the EEID data structure
+records the measurement after the heap and thread sections are added. This type
+of EEID enclave's SECS.Enclave_Size is determined by the base enclave.
+
+For EEID enclaves with heap/stack size and #TCS selected by the loader at
+enclave loading time, the loader adds heap pages, thread sections and the EEID
+page after the base enclave's code and initialized data are measured by the CPU.
+The `hash_state` records the measurement before the heap and thread sections are
+added. This type of of EEID enclave's Enclave_Size in `SECS` is fixed as a large
+value, for example 64GB.
+
+As discussed in the original EEID design doc, an EEID enclave with
+heap/stack size and #TCS selected by the loader at enclave load time needs to
+have one thread section in the base image to capture the expected `TCS.OENTRY`
+value in the base enclave measurement, which does complicate the base enclave
+measurement recreation algorithm used in EEID enclave attestation verification.
+This updated design addresses the need to capture the expected `TCS.OENTRY`
+value in a different way.
+
+`oe_sgx_enclave_image_info_t` is part of the initialized data of the SGX
+enclave, and covered by the base enclave measurement. The enclave loader will
+include the expected `TCS.OENTRY` value in `oe_sgx_enclave_image_info_t`. A new
+`entry_rva` member is added to `oe_sgx_enclave_image_info_t` definition:
+
+```C
+/* Extends oe_enclave_properties_header_t base type */
+typedef struct _oe_sgx_enclave_image_info_t
+{
+    uint64_t oeinfo_rva;
+    uint64_t oeinfo_size;
+    uint64_t reloc_rva;
+    uint64_t reloc_size;
+    uint64_t heap_rva; /* heap size is in header.sizesettings */
+    uint64_t enclave_size;
+    uint64_t entry_rva; /*from oeimage.entry_rva, must match TCS.OENTRY */
```

Comment:
I was thinking about this a bit more and I don't think this strategy works. The problem is that this implies that the enclave itself will validate that the measured `entry_rva` matches the TCS.OENTRY, but:
1. it requires a call-in to transfer control to the enclave, by which time a modified TCS.OENTRY would have already pre-empted the check. 
2. the enclave itself does not have read access to the contents of the TCS page, per PT_TCS definition. 

I'm not sure that you can get around having to measure and sign at least one TCS in the signing of the base enclave.

Reply:
Yeah, I also found that the TCS page isn't accessible, but I thought this would have to change for Icelake, because otherwise there's no way to TCSs that are dynamically added at runtime. Perhaps those pages would have to be added as a normal heap-type message and then the enclave changes it to `PT_TCS` from inside, a bit like EACCEPT'ing it but with the additional check?

I think it's fine if we keep the 0/0/1 settings as they are now, but it would of course be nice to change that to 0/0/0 if we find a way.

Reply:
The enclave code checks the measured entry_rva in `_or_sgx_enclave_image_info_t` against the value set by the loader in the eeid page (not agianst the value in TCS), before any additional content is loaded into the enclave. But I do realize that it requires further security analysis to understand whether EENTERing the enclave at an unexpected entry point through a TCS added by a malicious loader can bypass the checking logic somehow. I aslo think we can start with 0/0/1 settings, if the security analysis shows that the same attack will be detected for sure with the 0/0/1 settings.

EDMM does suport coverting regualr page to TCS. That's exactly how a TCS page will be added/acceptted.

Reply:
I discovered more potential attack techniques. The latest update to the design doc introduces an "eeid measurement context" page to be added to the base enclave measurement. The context page include all loader controlled/provided setting, outside of the SW convention, needed by the eeid enclave attestation verifier to recreate the HW reported `MRENCLAVE` and the base enclave's `SIGSTRUCT.ENCLAVEHASH`. If a malicious loader spoofs those settings, the base enclave measurement recalculated based on the spoofed settings will not match `SIGSTRUCT.ENCLAVEHASH`.